### PR TITLE
Disk backed storage service for local dev

### PR DESCRIPTION
**Pull request by mmihic**
Adds a storage service implementation backed by local disk, suitable for
use in development mode where you want data preserved across service
restarts. Works around the lack of a GCS emulator in the gcloud tools.
Supports multiple services pointing to the same local storage directory,
effectively emulating sharing GCS buckets across services.

Also adds more tests for the storage interfaces.

### CODE REVIEW

File: misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+import kotlin.streams.asSequence
+
+/**
+ * Implementation of [StorageRpc] that is backed by local disk storage. Useful for running
+ * in development mode on local machines, since there is no GCS emulator available. Files
+ * are stored with the generation version appended as an extension suffix, with a symlink
+ * existing for the latest generation. To preserve GCS atomicity semantics, progressive
+ * uploads are handled by storing the interim uploaded data in a temporary file based
+ * off the latest generation, then doing a rename to the new generation once the upload
+ * is complete.
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt]: love this

File: misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+ *
+ * The implementation uses file locks to support multiple local processes accessing the same
+ * storage directory. Whenever a blob is updated, the local store will acquire an exclusive
+ * lock on a corresponding lock file, releasing that lock when the update is complete (or when
+ * the process dies). It's slightly more complicated since we need to deal with the possibility
+ * that writer stops partway through without the process failing; in those cases we don't want to
+ * prevent subsequent updates from other processes.
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt]: I wonder if this is too much, and just atomic renaming is the only concurrency primitive we should have.

(Atomic renaming is pretty limited)

Comment (swankjesse) [reply misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt]: ... oh and that approach is totally broken if you want to do generation numbers. Never mind me.

File: misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+ *
+ * Read process
+ *  - acquire a read lock
+ *  - read the metadata for the blob and check constraints (including etag)
+ *  - read the raw bytes from the content file
+ *  - release the read lock
+ *
+ * Clients use etags to detect when a blob is updated while a progressive download is in place; the
+ * etag returned from a prior read is sent in subsequent reads. We simply use the generation number
+ * as the etag value.
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt]: that is marvelously simple

File: misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt [hunk 4]
```diff
@@ -0,0 +1,582 @@
...
+ *
+ */
+class LocalStorageRpc(root: Path, moshi: Moshi) : BaseCustomStorageRpc() {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt]: I think you can save the line below by defining moshi with a default value of `Moshi.Builder().build()`

File: misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt [hunk 5]
```diff
@@ -0,0 +1,582 @@
...
+    override fun create(
+            obj: StorageObject,
+            content: InputStream,
+            options: Map<StorageRpc.Option, *>
+    ): StorageObject? {
+        try {
+            val upload = beginUpload(obj, options)
+            var destOffset: Long = 0
+            val source = Okio.buffer(Okio.source(content))
+            source.forEachBlock(1024) { buffer, bytesRead ->
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt]: forEachBlock! Neat

File: misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+                }
+            }
+        } catch (e: IOException) {
+            throw StorageException(e)
+        }
+    }
+
+    private fun writeMetadata(blobId: BlobId, metadata: BlobMetadata) {
+        // NB(mmihic): Must be called with a write lock
+        val metadataJson = metadataAdapter.toJson(metadata).toByteArray()
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt]: ByteString.encodeUtf8() instead? I worked on mobile stuff that shipped everywhere and I’m still recovering from the default charset being relevant for calls like this

File: misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+        Files.createDirectories(metadataPath.parent)
+
+        return if (Files.exists(metadataPath)) {
+            val metadataContent = String(Files.readAllBytes(metadataPath))
+            metadataAdapter.fromJson(metadataContent)
+        } else null
+    }
+
+    private fun getContentSize(contentPath: Path) = try {
+        Files.newByteChannel(contentPath, READ).size()
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt]: need to close:
```
  Files.newByteChannel(contentPath, READ).use {
    return it.size()
  }
```

File: misk/src/main/kotlin/misk/io/PathExtensions.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package misk.io
+
+import java.nio.file.Files
+import java.nio.file.Path
+import kotlin.streams.asSequence
+
+/** @return all of the paths beneath this one, including nested paths */
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/io/PathExtensions.kt]: nice

File: misk/src/main/kotlin/misk/io/PathExtensions.kt [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+fun Path.listRecursively(includeDirs: Boolean = false): List<Path> {
+    val results = mutableListOf<Path>()
+    listRecursively(includeDirs, results)
+    return results.toList()
+}
+
+private fun Path.listRecursively(includeDirs: Boolean, results: MutableList<Path>) {
+    val (dirs, files) = Files.list(this).asSequence().partition { Files.isDirectory(it) }
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/io/PathExtensions.kt]: cute

File: misk/src/main/kotlin/misk/moshi/MoshiExtensions.kt [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package misk.moshi
+
+import com.squareup.moshi.Moshi
+
+inline fun <reified T> Moshi.adapter() = adapter(T::class.java)
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/moshi/MoshiExtensions.kt]: nice

File: misk/src/main/kotlin/misk/okio/OkioExtensions.kt [hunk 11]
```diff
@@ -0,0 +1,16 @@
+package misk.okio
+
+import okio.Buffer
+import okio.BufferedSource
+
+fun BufferedSource.forEachBlock(buffer: ByteArray, f: (buffer: ByteArray, bytesRead: Int) -> Unit) {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/okio/OkioExtensions.kt]: cool!

File: misk/testing/src/main/kotlin/misk/testing/TemporaryFolder.kt [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    /** Deletes all files and folders under the temporary folder */
+    fun delete() {
+        root.toFile().deleteRecursively()
+    }
+
+    /** @return a new folder with a randomly generated name */
+    fun newFolder(): Path = Files.createTempDirectory(root, "")
+
+    /** @return a new folder with the given name */
+    fun newFolder(name: String): Path = Files.createFile(root.resolve(name))
```
Comment (swankjesse) [inline misk/testing/src/main/kotlin/misk/testing/TemporaryFolder.kt]: createDirectory?

File: misk/testing/src/main/kotlin/misk/testing/TemporaryFolder.kt [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        context.store("tempFolder", tempFolder)
+
+        tempDir.toFile().deleteOnExit()
+
+        @Suppress("UNCHECKED_CAST")
+        val tempFolderProperty = context.requiredTestClass.kotlin.memberProperties.firstOrNull {
+            it is KMutableProperty<*> && it.returnType.isSubtypeOf(temporaryFolderType)
+        } as? KMutableProperty<TemporaryFolder>
+
+        tempFolderProperty?.setter?.call(context.requiredTestInstance, tempFolder)
```
Comment (swankjesse) [inline misk/testing/src/main/kotlin/misk/testing/TemporaryFolder.kt]: hmmm . . .  this is a little weird. I like the MiskTest injecting stuff; this is more bespoke.

Comment (swankjesse) [reply misk/testing/src/main/kotlin/misk/testing/TemporaryFolder.kt]: And JUnit 4 rules were better!

File: misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt (hunk 2 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt]: Also uploads are broken into chunks, so you need to store the chunks in a temp location until all of them are received, at which point you can do the atomic move.

File: misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt (hunk 4 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt]: oh yeah duh

File: misk/testing/src/main/kotlin/misk/testing/TemporaryFolder.kt (hunk 12 above)
Comment (mmihic) [reply misk/testing/src/main/kotlin/misk/testing/TemporaryFolder.kt]: woops

File: misk/testing/src/main/kotlin/misk/testing/TemporaryFolder.kt (hunk 13 above)
Comment (mmihic) [reply misk/testing/src/main/kotlin/misk/testing/TemporaryFolder.kt]: Agreed on Junit 4 rules.

I'll move into a `TemporaryFolderTestModule` which you can install with `MiskTest`.

I'm more and more leaning towards `MiskTest` just creating an injector, and having tests install modules which bind test callbacks for all test handling.

File: misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt (hunk 6 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt]: I think the case is a bit different - `toByteArray` explicitly uses UTF-8 as its default parameter.

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/to-byte-array.html

**Comment by mmihic** [comment]
Fast check mark would be good for now since this is dev-mode code; but a longer review later is always welcome.

**Pull request merged**